### PR TITLE
FIX: ensure type stability for missing cmaps in `set_cmap`

### PR DESCRIPTION
**Pull request by tacaswell**
## PR Summary

Previously this would raise a `ValueError`, the refactor caused it to raise
`KeyError`.   This inadvertently located a bad test in x-array, but restoring
the old behavior.

https://github.com/pydata/xarray/runs/8104117777?check_suite_focus=true#step:8:673

## PR Checklist

**Tests and Styling**
- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).

### CODE REVIEW

**Review by QuLogic — APPROVED** [review]

**Comment by QuLogic** [comment]
FYI, @mwaskom; not sure if you'll want to keep this compat bit: https://github.com/mwaskom/seaborn/pull/2971/files#diff-c6ea5a328fda7cdae50968540e9b71734bbf30b191ef2d9b710b60847cacb9edR208

File: lib/matplotlib/cm.py [hunk 1]
```diff
@@ -701,6 +701,6 @@ def _ensure_cmap(cmap):
     """
     if isinstance(cmap, colors.Colormap):
         return cmap
-    return mpl.colormaps[
-        cmap if cmap is not None else mpl.rcParams['image.cmap']
-    ]
+    cmap_name = cmap if cmap is not None else mpl.rcParams["image.cmap"]
+    _api.check_in_list(sorted(_colormaps), cmap=cmap_name)
```
Comment (greglucas) [inline lib/matplotlib/cm.py:705]: Do you want to put a comment in saying you added this check for error raising type stability so we don't remove it later? Does it need to be sorted or can it be `list(_colormaps)`?

**Review by greglucas — APPROVED** [review]
Just a minor nit, feel free to ignore and merge if you are fine with it currently.

File: lib/matplotlib/cm.py (hunk 1 above)
Comment (tacaswell) [reply lib/matplotlib/cm.py:705]: The input is used in formatting the error to tell the user what the allowed values are.  Given the number of color maps I think sorting is a usability win / kindness for the user to discover what the correct spelling of the color map they tried to ask for is.

Comment (tacaswell) [reply lib/matplotlib/cm.py:707]: I pushed a comment about why we are doing extra work.

**Pull request merged**